geomesa-kafka-bundle/geomesa-kafka-processors/src/main/scala/org/geomesa/nifi/processors/kafka/GetGeoMesaKafkaRecord.scala

```diff
@@ -139,95 +134,111 @@ class GetGeoMesaKafkaRecord extends AbstractProcessor {
+    val records = processor.ready.poll(pollTimeout, TimeUnit.MILLISECONDS)
+    if (records == null) {
+      context.`yield`()
+    } else {
+      val results = records.grouped(maxBatchSize).toSeq.map { group =>
+        var flowFile: FlowFile = null
+        try {
+          flowFile = session.create()
+          val result = WithClose(session.write(flowFile)) { out =>
+            WithClose(factory.createWriter(logger, schema, out, flowFile)) { writer =>
+              writer.beginRecordSet()
+              group.foreach { feature =>
+                writer.write(converter.convert(feature))
+              }
+              val result = writer.finishRecordSet()
+              WriteResult(result.getRecordCount, writer.getMimeType, result.getAttributes)
             }
-            val result = writer.finishRecordSet()
-            writer.flush()
-            WriteResult(result.getRecordCount, writer.getMimeType, result.getAttributes)
           }
-        }
-
-        if (result.count <= 0) {
-          logger.debug("Removing flow file, no records were written")
-          session.remove(flowFile)
-        } else {
           val attributes = new java.util.HashMap[String, String](result.attributes)
           attributes.put(CoreAttributes.MIME_TYPE.key, result.mimeType)
           attributes.put("record.count", String.valueOf(result.count))
-          session.transfer(session.putAllAttributes(flowFile, attributes), SuccessRelationship)
+          Success(session.putAllAttributes(flowFile, attributes))
+        } catch {
+          case NonFatal(e) =>
+            if (flowFile != null) {
+              Try(session.remove(flowFile)).failed.foreach(e.addSuppressed)
+            }
+            Failure(e)
         }
-      } catch {
-        case NonFatal(e) =>
-          logger.error("Error onTrigger:", e)
-          session.remove(flowFile)
       }
-    }
-  }
 
-  @tailrec
-  private def writeRemaining(writer: RecordSetWriter, timeout: Long, count: Int): Unit = {
-    if (timeout > System.currentTimeMillis()) {
-      val record = readNextRecord()
-      if (record != null) {
-        val count = writer.write(record).getRecordCount
-        if (count < maxBatchSize) {
-          writeRemaining(writer, timeout, count)
+      val errors = results.collect { case Failure(e) => e }
+      if (errors.nonEmpty) {
+        try {
+          val e = errors.head
+          errors.tail.foreach(e.addSuppressed)
+          logger.error("Error processing message batch:", e)
+          results.collect { case Success(f) => Try(session.remove(f)) }
+        } finally {
+          processor.done.put(false)
         }
-      } else if (count < minBatchSize) {
-        writeRemaining(writer, timeout, count)
+      } else {
+        var ok = true
+        results.foreach { case Success(f) =>
+          try {
+            if (f.getAttribute("record.count").toInt <= 0) {
+              logger.debug("Removing flow file, no records were written")
```

Comment:
What would happen if a flowfile writer failed/hung on a piece of data?  

Would a given batch be retried indefinitely?

Reply:
so errors in deserialization will be caught and logged, but won't be passed along to the client processor - see https://github.com/locationtech/geomesa/pull/2647/files#diff-db0118a973c17f07e1ee12bb8aeb0b5826e378e5c08b3dd7ab918a2aaae7ae14R63
if the client processor fails repeatedly on a piece of data for some reason, and it didn't have any logic to handle it (which this one doesn't), then yes it would flail repeatedly on the same batch. in this particular case, I'm not sure what would cause that to happen for one particular record but not others though... if you had a mis-configured recordSetWriter it seems possible, but then it seems like you would want to identify the issue and fix it rather than continuing on and dropping the data.